ykrt/src/mt.rs

```diff
@@ -342,13 +322,28 @@ impl MT {
     }
 
     /// Add a compilation job for `sir` to the global work queue.
-    fn queue_compile_job(&self, trace: IRTrace, mtx: Arc<Mutex<Option<Box<CompiledTrace>>>>) {
+    fn queue_compile_job(&self, trace: IRTrace, hl_ptr: *const HotLocation) {
+        let hl_ptr = hl_ptr as usize;
         let f = Box::new(move || {
             let code_ptr = trace.compile();
-            *mtx.lock() = Some(Box::new(CompiledTrace::new(code_ptr)));
-            // FIXME: although we've now put the compiled trace into the mutex, there's no
-            // guarantee that the Location for which we're compiling will ever be executed
-            // again. In such a case, the memory has, in essence, leaked.
+            let ct = Box::new(CompiledTrace::new(code_ptr));
+            let tmp_ls = LocationInner::new().with_hotlocation(hl_ptr as *mut HotLocation);
```

Comment:
This seems subtle. Can we add a comment explaining what is happening?

I *think* we use a temporary location with a copy of the heap pointer (whose contents may be changed between compilation starting and completing), so that the original location can still be dropped without the compilation thread being hosed. Right?

Reply:
Good point. Fixed in a77b95f.